File: Swift/프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.swift
import Foundation

struct Queue<T> {
    private var queue: [T] = []
    
    init(_ elements: [T]) {
        self.queue = elements
    }
    
    var isEmpty: Bool {
        return queue.isEmpty
    }
    
    var count: Int {
        return queue.count
    }
    
    func peek() -> T? {
        return queue.first
    }
    
    mutating func addLast(_ element: T) {
        queue.append(element)
    }
    
    mutating func pollFirst() -> T? {
        return queue.removeFirst()
    }
}

func solution(_ bridge_length:Int, _ weight:Int, _ truck_weights:[Int]) -> Int {
    var answer: Int = 0
    var bridgeWeight: Int = 0
    var trucks: Queue<Int> = Queue<Int>(truck_weights)
    var destTime: Queue<Int> = Queue<Int>([])
    var passingTruck: Queue<Int> = Queue<Int>([])
    var completedPassingTruck: Int = 0
    
    while (completedPassingTruck < truck_weights.count) {
         if destTime.peek() == answer {
            bridgeWeight -= passingTruck.pollFirst()!
            _ = destTime.pollFirst()
            completedPassingTruck += 1
        }
        
        if let truckPeek = trucks.peek(), bridgeWeight+truckPeek <= weight, passingTruck.count < bridge_length {
            destTime.addLast(answer+bridge_length)
            let truck = trucks.pollFirst()!
            passingTruck.addLast(truck)
            bridgeWeight += truck
        }
        
        answer += 1
    }
    
    return answer
}

File: Java/프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.ArrayDeque;
import java.util.Arrays;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        ArrayDeque<Integer> waitingTruck = new ArrayDeque<>();
        for (int w : truck_weights) {
            waitingTruck.add(w);
        }
        
        ArrayDeque<Integer> outBridgeTime = new ArrayDeque<Integer>();
        ArrayDeque<Integer> onBridge = new ArrayDeque<Integer>();
        int currentWeight = 0;
        int currentTime = 0;
        int destinationCount = 0;
        
        while (destinationCount < truck_weights.length) {
            if (!onBridge.isEmpty() && outBridgeTime.peekFirst() == currentTime) {
                outBridgeTime.pollFirst();
                currentWeight -= onBridge.pollFirst();
                destinationCount++;
            }
            
            if (!waitingTruck.isEmpty() && currentWeight+waitingTruck.peekFirst() <= weight && onBridge.size() < bridge_length) {
                outBridgeTime.addLast(currentTime+bridge_length);
                int truck = waitingTruck.pollFirst();
                onBridge.addLast(truck);
                currentWeight += truck;
            }
            
            currentTime++;
        }

        return currentTime;
    }
}